### PR TITLE
Add support jpeg Exif.Image.Orientation (fix #2525)

### LINKED ISSUE #2525: Support jpeg Exif.Image.Orientation tag
Aseprite doesn't appear to support the Exif.Image.Orientation tag. I took a photo of an object with my iPhone in order to import it into Aseprite as a reference layer. When opening the photo in Apple's Preview app, it shows right-side up, but in Aseprite, it's upside-down. In Preview, the Info window shows "Orientation: 3 (Rotated 180°)". Or from the command line:

```
$ exifprint IMG_7288.JPG | grep Orientation
Exif.Image.Orientation                       0x0112 Short       1  3
```

(`exifprint` is part of [exiv2](https://www.exiv2.org).)

I suspect the Orientation depends on which way I had rotated the phone when I took the photo.

### Aseprite and System version

* Aseprite version: 1.2.25 compiled from source
* System: macOS High Sierra 10.13.6

### PR DESCRIPTION
**Pull request by Gasparoken**
Added tinyexif submodule also.

Pending adjust some CMakeLists.txt files

fix #2525

### CODE REVIEW

**Comment by dacap** [comment]
The commit doesn't include the new submodule. Submodules are not added manually in the `.gitmodules` file, they must be added with the `git submodule` command (which creates a special metadata in the commit referencing a specific commit of the submodule).

You will have to revert the `.gitsubmodules` changes, remove third_party/tinyexif directory, and run:

    git submodule add https://github.com/aseprite/tinyexif.git third_party/tinyexif

Then git add the changes and amend your commit.

**Comment by dacap** [comment]
@Gasparoken I just saw that in the commit you've pushed b701e3eca8f9afe12a3c2d2d25e6445dc95dda4c I'm in the author field. This is an honest question: how did you achieve that? because it's quite complex to achieve that. You have to explicitly use the --author parameter, so I'm not sure if the tool you are using is making things more difficult that they should for you.

What Git client are you using and how did you create this commit?

**Comment by Gasparoken** [comment]
> @Gasparoken I just saw that in the commit you've pushed [b701e3e](https://github.com/aseprite/aseprite/commit/b701e3eca8f9afe12a3c2d2d25e6445dc95dda4c) I'm in the author field. This is an honest question: how did you achieve that? because it's quite complex to achieve that. You have to explicitly use the --author parameter, so I'm not sure if the tool you are using is making things more difficult that they should for you.
> 
> What Git client are you using and how did you create this commit?

My honest answer: VSCode, I never used the 'author' command.

**Comment by dacap** [comment]
It's super strange what VSCode did there, is like it copied the author and date of the previous commit to amend your commit. I'll try to reproduce something like this and bring some help.

**Comment by dacap** [comment]
At least using "Commit (Amend)" in these cases (where you have to adjust your latest commit changes) works as expected: (the author doesn't change)

image

Do you remember if that is the option you've used?

**Comment by Gasparoken** [comment]
I used the option, Commit > Commit Staged (Amend).

**Comment by dacap** [comment]
> I used the option, Commit > Commit Staged (Amend).

Did you close VSCode from yesterday? If that is not the case, it would great if you can send me your Git output to david@igara.com:

image

**Comment by Gasparoken** [comment]
Hi @dacap. I need some advice on Linux compilation. I'm not sure what's wrong during cmake configuration.

File: third_party/CMakeLists.txt [hunk 1]
```diff
@@ -58,6 +58,15 @@ if(NOT USE_SHARED_TINYXML)
   add_subdirectory(tinyxml2)
 endif()
 
+# Jpeg exif fields
+if(NOT USE_SHARED_TINYEXIF)
+  set(TinyEXIF_BUILD_TESTING OFF CACHE BOOL "Build tests for tinyexif")
+  add_subdirectory(TinyEXIF)
```
Comment (dacap) [inline third_party/CMakeLists.txt:64]: Here the subdirectory is added as "TinyEXIF" but the submodule cloned in "third_party/tinyexif". You should clone add the submodule as it is: "TinyEXIF" in "third_party/TinyEXIF", rename the repository etc. Probably you will need to remove the submodule and add again correctly in this commit.

Exactly as what is done with IXWebSocket and https://github.com/aseprite/IXWebSocket

**Comment by dacap** [comment]
I've just added the TinyEXIF module in:

https://github.com/aseprite/TinyEXIF/commit/2283d8094e42e6b6cef00d1d31ecadfe0d9e1f29
https://github.com/dacap/aseprite/commit/c83dd16a43fd52ead709f0c9e78e7d7acb54dac1

These changes are already in `main` branch so you can continue this PR from there.

File: src/app/file/jpeg_format.cpp [hunk 2]
```diff
@@ -177,12 +178,38 @@ bool JpegFormat::onLoad(FileOp* fop)
...
 
+  // Get orientation:
+  TinyEXIF::EXIFInfo info;
+  int orientation = 0;
+#if LAF_WINDOWS
+  const int tempBufSize = (int)std::ftell(file);
+  rewind(file);
+  char *tempBuf = (char *)malloc(sizeof(char) * tempBufSize);
+  if (!tempBuf) {
+    std::free(tempBuf);
```
Comment (martincapello) [inline src/app/file/jpeg_format.cpp:189]: If tempBuf is null there is no point in trying to deallocate it. Anyway, see my other comment.

File: src/app/file/jpeg_format.cpp [hunk 3]
```diff
@@ -177,12 +178,38 @@ bool JpegFormat::onLoad(FileOp* fop)
...
+  orientation = info.Orientation;
+  std::free(tempBuf);
+#elif LAF_MACOS
+  info.parseFrom(file->_bf._base, file->_bf._size);
+  orientation = info.Orientation;
+#elif LAF_LINUX
+  const int fileSize = std::ftell(file);
+  info.parseFrom((uint8_t*)file->_IO_read_base, fileSize);
+  orientation = info.Orientation;
+#endif
```
Comment (martincapello) [inline src/app/file/jpeg_format.cpp:204]: I think we can avoid using platform specific code if we put something like the following in line 144:
```c++
  int orientation = 0;
  {
    std::ifstream istream(fop->filename(), std::ifstream::in | std::ifstream::binary);
    // Get EXIF information:
    TinyEXIF::EXIFInfo info(istream);
    if (info.Fields)
      orientation = info.Orientation;
  }
```

**Review by martincapello — COMMENTED** [review]
Works great! I've tested all orientations in landscape and portrait mode and it works as expected. Just left one comment to avoid platform specific code when reading the the EXIF info, the rest is just fine for me.

**Comment by dacap** [comment]
It works great, I've just tested with these images: https://github.com/ianare/exif-samples/tree/master/jpg/orientation

**Pull request merged**